### ISSUE TITLE
Proposal : allow ignorable <xsl:div> wrapper for documentation or organize the code

### ISSUE THREAD

**Issue by mricaud** [qt4cg/qtspecs#84]
Hi,

It's a long time I'm missing a way to organize XSLT code. Using [enclosed mode](https://qt4cg.org/branch/master/xslt-40/Overview-diff.html#enclosed-modes) will help a lot, but it's not its main purpose and it's not enought I guess to be able : 
- Group templates or function that go together (according to the author)
- easily comments blocks of code for debugging purpose
- add documentation on any XSLT elements : not only top level elements as with oxygen "xd" elements
- add foreign xml structures that can help for static analysing of the code (ex : informations to help with xslt schematron validation, that need autocompletion with a specific xml schema, that means processing instructions are not engough here)

What about a `xsl:div` element (for division), this is a well known element's name, used in HTML but also Relax NG.
That element might have a `process-content` attribute with 2 possible values:
- `true` (default): to say the content should be "applied" as `xsl:div` might be nested  with different process-content attribute values
- `false`: to say the content has to be completely skipped at compilation

Example:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet 
  xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
  xmlns:xs="http://www.w3.org/2001/XMLSchema"
  xmlns:xd="http://www.oxygenxml.com/ns/doc/xsl"
  xmlns:local="local"
  xmlns:xslq="https://github.com/mricaud/xsl-quality"
  xmlns:a="http://my-annotations.org"
  version="4.0">
  
  <xsl:div>
    <xsl:div process-content="false">
      <p>This block is about "foo"</p>
    </xsl:div>
    
    <xsl:function name="local:has-foo-child" as="xs:boolean">
      <xsl:param name="e">
        <xsl:div process-content="false"><xd:doc>Any elements</xd:doc></xsl:div>
      </xsl:param>
      <xsl:sequence select="exists($e/foo)"/>
    </xsl:function>
    
    <xsl:template match="foo">
        <xsl:div process-content="false">
           <xslq:schematron ignore-rule="mode-name-must-be-namepace-prefixed"/>
         <xsl:div>
        <xsl:value-of select="normalize-space(.)"><xsl:div process-content="false"><xd:doc>Normalization is needed here</xd:doc></xsl:div>
    </xsl:template>
    
  </xsl:div>
  
</xsl:stylesheet>
```

Writing this example let me see that it's a bit verbose. Another proposal would be to declare a set of namespaces that are to be ignored at compilation time, wether by skipping it or by apply what's inside of it. 
The same example would give something like this:

```xml
<xsl:stylesheet 
  xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
  xmlns:xs="http://www.w3.org/2001/XMLSchema"
  xmlns:xd="http://www.oxygenxml.com/ns/doc/xsl"
  xmlns:local="local"
  xmlns:xslq="https://github.com/mricaud/xsl-quality"
  xmlns:a="http://my-annotations.org"
  version="4.0">
  
  <xsl:ignore-namespaces select="map{
    'http://www.oxygenxml.com/ns/doc/xsl' : 'skip', 
    'http://my-annotations.org' : 'apply',
    'https://github.com/mricaud/xsl-quality' : 'skip'
    }"/>
    
  
  <a:div label="This block is about foo">
    
    <xsl:function name="local:has-foo-child" as="xs:boolean">
      <xsl:param name="e"><xd:doc>Any elements</xd:doc></xsl:param>
      <xsl:sequence select="exists($e/foo)"/>
    </xsl:function>
    
    <xsl:template match="foo" mode="bar">
      <xslq:schematron ignore-rule="mode-name-must-be-namepace-prefixed"/>
      <xsl:value-of select="normalize-space(.)" xd:doc="normalization is needed here"/>
    </xsl:template>
    
  </a:div>
  
</xsl:stylesheet>
```

Well, there are probably multiple ways to achieve this need.

I would be really happy to have such possiblilities, but I don't know if some of you have the same need.

Thanks for any comments / replies / ideas / feedbacks

**Comment by gimsieke** [comment]
I suggest that the attribute be [`@use-when`](https://www.w3.org/TR/xslt-30/#conditional-inclusion) instead of `@process-content` because its semantics is established (static expression). Example:

```xml
  <xsl:param name="foo" static="yes" as="xs:string?"/>
  <xsl:div use-when="$foo = 'bar'">
    <xsl:template match="/*" priority="2">
      …
    </xsl:template>
    <xsl:template match="@xml:lang"/>
  </xsl:div>
```

Right now you can attach these conditions to individual elements, which I personally think is sufficient. 

An advantage of `@use-when` is that it can hold any expression that can be statically computed, not just 'true' or 'false'. And you can do your own grouping by using static parameters with different names and values. The only drawback that I see that the grouping is virtual, that is, by the computed `@use-when` value; there is no `xsl:div` element to group contiguous elements.

Another advantage of `@use-when` is that you can attach its `@xsl:use-when` variant to other elements than those in the `xsl:` namespace.

```xml
  <xsl:template name="test">
    <xsl:param name="p">
      <xsl:sequence use-when="$foo = 'bar'" select="'bar-default-value'"/>
      <xsl:sequence use-when="not($foo = 'bar')" select="'default-value'"/>
    </xsl:param>
      …
  </xsl:template>
```
I’m not strictly against grouping contiguous stylesheet elements by `xsl:div` though if more users find it useful and if isn’t too hard to implement for XSLT processor vendors.

**Comment by mricaud** [comment]
Thanks for you response @gimsieke, I didn't know `@use-when` could be used on non-xsl elements, that's interresting. 
But in the cases of organizing / annotating XSLT code, it would probably be a missused of `@use-when` : the value would always be "false()" whereas this attribute is done to activate **or not** at compilation time. It would work but I would see it like a kind of "hack" instead of something intended to the real purpose.
Another drawback is about validation : using a wrapper like `xsl:div` or adding extra annotation within `xsl:function` or `xsl:template` might raise XSLT validation errors. For example if I add an annotation before xsl:param in a function, etc.
This sould be possible at IDE level to correct this, but this make the validation process much more complex ?

Another way to achieve this would be to use PI or comments, but then one couldn't easily enjoy XML capabilities to generate documentation.

I have a project to statically analyze XSLT with schematron, it helps to avoid errors and consolidate the code with good practice.
But because one can not set global rules that are true for each developer or each XSLT, I set up a customization annotation to be added within the XSLT. At top level, no problem I can use regular xml element (xslq:conf) with a specific schema. Autocompletion (really usefull here) works fine in Oxygen. But inside function/template I had to make it different with attributes.
Here is an example : https://github.com/mricaud/xslt-quality/blob/master/src/test/sample-conf.xsl

**Comment by mricaud** [comment]
I just realized this issue is quite the same as issue #19, thanks @liamquin and sorry for not seeing it before

**Comment by yamahito** [comment]
Is it exactly the same?  I thought that Liam's suggestion was not just for the wrapper to be ignored, but all of its contents: I understood that your suggestion was for the wrapper to be ignored, but NOT its contents - is that right?

**Comment by mricaud** [comment]
Hi @yamahito, I should actually change the subject of this issue, as I finally propose to set a list of namespaces to be ignored like Liam. But yes you're right, the difference is that I consider 2 behaviours : ignore the content or go into it.
The main difference is about this xsl:div wrapper : 
- adding it and keep Liam's proposal would make it good
- using enclosed mode is also an alternativ but less possibilities of grouping than xsl:div 
- Instead of defining xsl:div mixing Liam's proposal with the ability to add the "skipping behaviour" ?

**Comment by michaelhkay** [comment]
This can now be closed following adoption of PR #353 (xsl:note)
